module/applications/advancement/item-choice-flow.mjs

```diff
@@ -0,0 +1,178 @@
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle deleting a dropped item.
+   * @param {Event} event  The originating click event.
+   * @protected
+   */
+  async _onItemDelete(event) {
+    event.preventDefault();
+    const uuidToDelete = event.currentTarget.closest(".item-name")?.querySelector("input")?.name;
+    if ( !uuidToDelete ) return;
+    this.dropped.findSplice(i => i.uuid === uuidToDelete);
+    this.selected.delete(uuidToDelete);
+    this.render();
+  }
+
+  /* -------------------------------------------- */
+
+  /** @inheritdoc */
+  async _onDrop(event) {
+    if ( this.selected.size >= this.advancement.configuration.choices[this.level] ) return false;
+
+    // Try to extract the data
+    let data;
+    try {
+      data = JSON.parse(event.dataTransfer.getData("text/plain"));
+    } catch(err) {
+      return false;
+    }
+
+    if ( data.type !== "Item" ) return false;
+    const item = await Item.implementation.fromDropData(data);
+
+    try {
+      this.advancement._verifyItemType(item);
+    } catch(err) {
+      return ui.notifications.error(err.message);
+    }
+
+    // If the item is already been marked as selected, no need to go further
+    if ( this.selected.has(item.uuid) ) return false;
+
+    // Check to ensure the dropped item hasn't been selected at a lower level
+    for ( const [level, data] of Object.entries(this.advancement.value) ) {
+      if ( level >= this.level ) continue;
+      if ( Object.values(data).includes(item.uuid) ) return;
+    }
+
+    // If spell level is restricted to available level, ensure the spell is of the appropriate level
+    const spellLevel = this.advancement.configuration.spellLevel;
+    if ( (this.advancement.configuration.type === "spell") && spellLevel === "available" ) {
+      const maxSlot = this._maxSpellSlotLevel();
+      if ( item.system.level > maxSlot ) return ui.notifications.error(game.i18n.format(
+        "DND5E.AdvancementItemChoiceSpellLevelAvailableWarning", { level: CONFIG.DND5E.spellLevels[maxSlot] }
+      ));
+    }
+
+    // Mark the item as selected
+    this.selected.add(item.uuid);
+
+    // If the item doesn't already exist in the pool, add it
+    if ( !this.pool.find(i => i.uuid === item.uuid) ) {
+      this.dropped.push(item);
+      item.dropped = true;
+    }
+
+    this.render();
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Determine the maximum spell slot level for the actor to which this advancement is being applied.
+   * @returns {number}
+   */
+  _maxSpellSlotLevel() {
+    const largestSlot = Object.entries(this.advancement.actor.system.spells).reduce((slot, [key, data]) => {
+      if ( data.max === 0 ) return slot;
+      const level = parseInt(key.replace("spell", ""));
+      if ( !Number.isNaN(level) && level > slot ) return level;
+      return slot;
+    }, -1);
+    return Math.max(this.advancement.actor.system.spells.pact.level, largestSlot);
```

Comment:
I think this will require two code paths. For cases where this Advancement is on a class or subclass, we can use the highest level the class can cast, but if it is on some other item like a background we'll use the max level available across all classes.

Reply:
Is there any precedent for a spell item grant from a non-class (like a race or feat) that is limited by your maximum spell slot level? Ritual Caster has its own limit for example (half your character level rounded up), but I don't think there's any general rule. Should there be no limit if there isn't an associated class or subclass then maybe? I'm not really sure.